staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types_discovery.go

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// APIGroupDiscoveryList is a resource containing a list of APIGroupDiscoveries.
+// The shape is chosen to be very similar to "normal" api endpoints and work well with our type system
+// This is what is returned from the /discovery/v2 endpoint.
+type APIGroupDiscoveryList struct {
+	TypeMeta `json:",inline"`
+	// ResourceVersion will not be set, because this does not have a replayable ordering among multiple apiservers.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Groups is the list of groups for discovery.
+	// +listType=map
+	// +listMapKey=name
+	Groups []APIGroupDiscovery `json:"groups" protobuf:"bytes,2,rep,name=groups"`
```

Comment:
We have a defined behavior for doing List, and it's for this to be called `Items`.  I'm not comfortable breaking that convention.  Either discovery/v1 returns a singular resource, which has nested arrays, or we return a list of items, in which case we are saying we want to support regular LIST semantics.  This type can only be called a "*List" if we are using list semantics, which we aren't.  So this should be called APIGroupsDiscovery or something similar as a singleton resource (if we're doing singletons).

That is also locking us into a path where we can't do watch the "normal" way (normally watch requires a LIST verb), which means we're making up a custom watch behavior if we introduce watching.  So the decision about WATCH is pretty important to the design of this.

Reply:
That's a good point. Updated the list to be called `Items`

Reply:
> If we go with a singular resource, it's locking us into a path where we can't do watch the "normal" way

We can (and probably should) place a new resource in the ordinary place if/when we want to support watch. I don't think it's critical that this data type match our regular resources since it is not a regular resource.

Reply:
Wait, so if we're gonna do it for watch.... why don't we just start with that?  Did you assess that as too complex?  Wouldn't that mean I have to do discovery to find the watchable discovery resource?

In general we've held that consistency of API behavior is more important than allowing special cases.  Consistency reduces friction for users to understand what they can do.  These resources share our serialization format, so there's some expectation of consistency at least for top level objects in serialization, and `List has Items` is a standard expectation (and makes various client helpers work correctly).

Reply:
Making this watchable doesn't solve the problem the team is trying to solve (kubelet makes many requests on startup).

Discovery must always be an exception since clients must go here before they know to go anywhere else. That means it won't be mechanically consistent no matter what else we do.

The consumer of discovery is not normal users. Normal users write controllers or use kubectl. The people who use discovery (i.e. program against it) are client library authors. I just don't think partial consistency here really helps many people, and it may be anti-help since we can't be perfectly consistent.

The things I think are important about this is that it should be easy to convert to/from a hypothetical watchable resource. And it should be easy to convert to/from the former version.

Reply:
> Discovery must always be an exception since clients must go here before they know to go anywhere else. That means it won't be mechanically consistent no matter what else we do.

I don't understand how that is true, lots of our clients go directly to APIs without performing discovery.  A generic client can go to a discovery endpoint that looks arbitrary (/discovery/X is arbitrary).  Perhaps you're saying - mechanical consistency of the URl / pattern for discovery SHOULD be different today from regular resources, I'm just trying to understand the "why".

> I just don't think partial consistency here really helps many people, and it may be anti-help since we can't be perfectly consistent.

I guess I'm asking because I didn't see an attempt to be consistent, or even the consideration.  I just may have missed it in the KEP or the discussion, in which case just registering it in the "alternate considered" would be sufficient.  I.e. if you had pitched in the kep "discovery would be done by calling LIST on /apis/v1/discovery.k8s.io/apiresources" would anyone have even noticed?  

> The things I think are important about this is that it should be easy to convert to/from a hypothetical watchable resource. And it should be easy to convert to/from the former version.

Agreed.  And "all things that are possible in the former version are available in the current version" is the specific sub point I brought up below, which is our default constraint unless we choose not to, and I didn't see us choosing not to.

But that being said, we're proposing some new magic type of resource, but I didn't see an alternate reasoning that excluded "just make this look like any other resource in our namespace" which is how we'd treat any other api.  You don't need to do discovery to query `/apis/v1/discovery.k8s.io/apiresources` first instead of `/discovery/v1` or `/apis` - the last two are arbitrary, the first one is arbitrary AND consistent.

Reply:
Resolving this as it is now covered in the KEP. We will be reusing the `/apis` and `/api`endpoints, watch is outside the scope of this design.